tephra-core/src/main/java/org/apache/tephra/distributed/TransactionService.java

```diff
@@ -121,15 +121,15 @@ public void failed(State from, Throwable failure) {
       @Override
       public void follower() {
         ListenableFuture<State> stopFuture = null;
+        if (pruningService != null && pruningService.isRunning()) {
+          // Wait for pruning service to stop after un-registering from discovery
+          stopFuture = pruningService.stop();
+        }
         // First stop the transaction server as un-registering from discovery can block sometimes.
         // That can lead to multiple transaction servers being active at the same time.
         if (server != null && server.isRunning()) {
           server.stopAndWait();
         }
-        if (pruningService != null && pruningService.isRunning()) {
-          // Wait for pruning service to stop after un-registering from discovery
-          stopFuture = pruningService.stop();
-        }
         undoRegister();
 
         if (stopFuture != null) {
```

Comment:
It would be good to add a timeout to the `Futures.getUnchecked(stopFuture)` in the next line, so that we don't block forever.

Reply:
pruningService.shutdown() itself has a 30 second timeout. So this future is guaranteed to return, and we don't need a timeout here.